### PR TITLE
Refactor DIMS sum adducts

### PR DESCRIPTION
**Pull request by mraves2**
DIMS pipeline refactor SumAdducts

The goal was to make this step of the pipeline faster, to tidy up the main script and sum_adducts function and to add unit tests.

main script: CustomModules/DIMS/SumAdducts.R

- moved function sum_adducts out of main script into Utils/sum_adducts.R
- removed unnecessary input parameters for function sum_adducts
- added docstrings to function sum_adducts
- writing output to file moved to main script
- function sum_adducts optimized by taking two steps out of while loop
- unit test in tests/testthat/test_sum_adducts.R

Questions about unit tests; all feedback is welcome.
- is it better to create test objects test_peakgroup_list and test_hmdb_main_part in script or to save them to files and read them in?
- is it necessary to set working directory in testthat.R (at least until functions are in package)?
- are there other tests that need to be added?
  - adducts <- c(1,2) is for positive scan mode, negative scan mode is even simpler (c(1)) and does not need to be tested imho
  - z_score = 0 is only for research, not diagnostics.
  
Code can be found in /hpc/dbg_mz/development/DIMS_refactor_SumAdducts

### CODE REVIEW

File: DIMS/Utils/sum_adducts.R [hunk 1]
```diff
@@ -0,0 +1,73 @@
+sum_adducts <- function(peakgroup_list, hmdb_part, adducts, z_score) {
+  #' Sum intensities for different adducts of the same metabolite
+  #'
+  #' @param peakgroup_list: Peak group list (matrix)
+  #' @param hmdb_part: List of metabolites, part of the HMDB (matrix)
```
Comment (Jorisvansteenbrugge) [inline DIMS/Utils/sum_adducts.R:5]: Perhaps `hmdb_subset` (instead of `hmdb_part`) would be a bit more descriptive variable name? 
I am also wondering if this variable is indeed a part/subset of the HMDB (matrix), or if the format is different? That is not entirely clear to me. The docstring mentions it being a list of metabolites, while HMDB seems to be a matrix.

File: DIMS/Utils/sum_adducts.R [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+  #' @param adducts: List of adducts (array of integers)
+  #' @param z_score: Value indicating whether Z-scores have been calculated (integer)
+  #'
+  #' @return adductsum: peak group list with summed intensities (matrix)
+  hmdb_codes <- rownames(hmdb_part)
```
Comment (Jorisvansteenbrugge) [inline DIMS/Utils/sum_adducts.R:10]: This ties in with my previous comment, `hmdb_part` is referred to as a list in the documentation, however, I believe rownames(list) returns `NULL`. So my guess would be that `hmdb_part` is a matrix, rather than a list?

File: DIMS/Utils/sum_adducts.R [hunk 3]
```diff
@@ -0,0 +1,73 @@
+sum_adducts <- function(peakgroup_list, hmdb_part, adducts, z_score) {
+  #' Sum intensities for different adducts of the same metabolite
+  #'
+  #' @param peakgroup_list: Peak group list (matrix)
+  #' @param hmdb_part: List of metabolites, part of the HMDB (matrix)
+  #' @param adducts: List of adducts (array of integers)
```
Comment (Jorisvansteenbrugge) [inline DIMS/Utils/sum_adducts.R:6]: Is adducts actually a `List` data type? The documentation describes it as both `List` and `array`, which are two different datatypes. When looking at the rest of the script `adducts` appears to be neither, but looks like a vector.
For reference,
- vector -> May contain multiple elements of the same `atomic` type (logical, integer, character, etc.)
- array -> a multidimensional vector with elements of the same `atomic` type
- list -> May contain multiple elements of different types (integers, characters, but also vectors, matrices, etc.)

I think correctly naming the datatype will improve clarity :)

File: DIMS/Utils/sum_adducts.R [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+    int_cols <- c(int_cols_start:int_cols_end)
+  }
+  
+  # initialize
+  names <- NULL
+  adductsum <- NULL
+  names_long <- NULL
+  
+  # find adducts of each metabolite and sum the intensities
+  if (length(hmdb_codes) != 0) {
```
Comment (Jorisvansteenbrugge) [inline DIMS/Utils/sum_adducts.R:46]: A suggestion to improve the readability (and reduce complexity) of this part is to use a [guard clause](https://deviq.com/design-patterns/guard-clause), instead of nesting the main logic inside the if statement.

So instead of writing:
```
if (length(hmdb_codes) != 0) {

 //the rest of the code

}
```

you could write it like:
```
if (length(hmdb_codes) != 0) {
    return(adductsum) # which is NULL in that case
}

// the rest of the code
```

This reduces the use of unnecessary indents.

File: DIMS/tests/testthat.R [hunk 5]
```diff
@@ -0,0 +1,16 @@
...
+library(testthat)
+
+# source all functions for running Outrider
+source("/Users/mraves2/Development/DIMS_refactor_SumAdducts/DIMS/CustomModules/DIMS/Utils/sum_adducts.R")
+
+# temporary (files are local):
+# setwd("/Users/mraves2/Genomics/RNAseq_outrider/tests")
+
+# set working directory to location of testthat.R file
+location_current_file <- rstudioapi::getSourceEditorContext()$path
```
Comment (Jorisvansteenbrugge) [inline DIMS/tests/testthat.R:11]: I would steer clear of using the rstudio api, otherwise the test will break when running on the command line, or when setting up automated testing in the future.

File: DIMS/tests/testthat.R [hunk 6]
```diff
@@ -0,0 +1,16 @@
+# Run all unit tests on sum_adducts
+library(testthat)
+
+# source all functions for running Outrider
+source("/Users/mraves2/Development/DIMS_refactor_SumAdducts/DIMS/CustomModules/DIMS/Utils/sum_adducts.R")
```
Comment (Jorisvansteenbrugge) [inline DIMS/tests/testthat.R:5]: I would try to make this a relative path or the test will not work anymore if this repository is checked out in a different location

**Review by Jorisvansteenbrugge — CHANGES_REQUESTED** [review]
Nice work, I have a few main comments in addition to the specific ones.

1. You moved a function with the name `sum_adducts` from DIMS/SumAdducts.R to DIMS/Utils/sum_adducts.R. This is a bit confusing because now we have two files more or less called 'sum adducts.R' but only one of them contains that function. I think It would be clearer to change the name of one of them to something else.

Regarding your questions about the unit tests:
- "is it better to create test objects test_peakgroup_list and test_hmdb_main_part in script or to save them to files and read them in?"
 **Creating them at runtime (like you have now) does improve reproducibility, but might be less performant if the dataframes are very large. Depending on that you could decide to store them in a file, and parse those in the test file. **
 - "is it necessary to set working directory in testthat.R (at least until functions are in package)?"
  **I think so, but I am not 100% sure**
 - "are there other tests that need to be added?"
  **Ideally you would add a test for each situation that might occur. Right now you wrote a test that checks if the output of a test dataset is as expected, but you also could write a test for a scenario when the data is malformed, or empty (so the edge cases) etc.**

File: DIMS/Utils/sum_adducts.R (hunk 1 above)
Comment (mraves2) [reply DIMS/Utils/sum_adducts.R:5]: It is indeed a subset of the HMDB but since we're so used to calling it a part of the HMDB, I will  continue with the current name for now. In a future refactor of the scripts where hmdb_parts are generated, I will rename to hmdb_subset.
Docstring modified.

File: DIMS/Utils/sum_adducts.R (hunk 2 above)
Comment (mraves2) [reply DIMS/Utils/sum_adducts.R:10]: Correct. hmdb_part is a matrix with a row for each metabolite.

File: DIMS/Utils/sum_adducts.R (hunk 3 above)
Comment (mraves2) [reply DIMS/Utils/sum_adducts.R:6]: Docstring description modified to vector.

File: DIMS/Utils/sum_adducts.R (hunk 4 above)
Comment (mraves2) [reply DIMS/Utils/sum_adducts.R:46]: I agree that the code would become cleaner, but the construction you suggest would lead to two return statements in one function, which can be confusing. Still, depending on the outcome of the if-statement, only one of the return statements will be used. I will incorporate your suggestion (I assume the if-statement should read:
if (length(hmdb_codes) == 0) {
    return(adductsum)
}

File: DIMS/tests/testthat.R (hunk 5 above)
Comment (mraves2) [reply DIMS/tests/testthat.R:11]: References to current location of file have been removed. How location of files should be referenced depends on GitHub Actions, to be implemented at a later stage.

File: DIMS/tests/testthat.R (hunk 6 above)
Comment (mraves2) [reply DIMS/tests/testthat.R:5]: Path rewritten as "../Utils/sum_intensities_adducts.R"
File Utils/test_sum_adducts.R renamed to Utils/test_sum_intensities_adducts.R in accordance to renaming of function that is being tested.

**Review by Jorisvansteenbrugge — APPROVED** [review]
Changes look good to me!
